File: services/api/src/main/java/com/launchly/release/controllers/ReleaseController.java
package com.launchly.release.controllers;

import com.launchly.common.security.AuthContext;
import com.launchly.release.dto.GateExemptionRequest;
import com.launchly.release.dto.ReleaseRequest;
import com.launchly.release.dto.ReleaseResponse;
import com.launchly.release.entities.GateExemption;
import com.launchly.release.services.GateCheckResult;
import com.launchly.release.services.ReleaseService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/projects/{projectId}/releases")
public class ReleaseController {
    private final ReleaseService releaseService;

    public ReleaseController(ReleaseService releaseService) {
        this.releaseService = releaseService;
    }

    @PostMapping
    public ResponseEntity<ReleaseResponse> create(@PathVariable String projectId,
                                                   @Valid @RequestBody ReleaseRequest request) {
        return ResponseEntity.ok(releaseService.createRelease(projectId, request, AuthContext.userId()));
    }

    @GetMapping
    public ResponseEntity<List<ReleaseResponse>> list(@PathVariable String projectId) {
        return ResponseEntity.ok(releaseService.listReleases(projectId));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ReleaseResponse> get(@PathVariable String projectId, @PathVariable String id) {
        return ResponseEntity.ok(releaseService.getRelease(id));
    }

    @GetMapping("/{id}/gates")
    public ResponseEntity<GateCheckResult> gates(@PathVariable String projectId, @PathVariable String id) {
        return ResponseEntity.ok(releaseService.getGateStatus(id));
    }

    @PutMapping("/{id}/publish")
    public ResponseEntity<ReleaseResponse> publish(@PathVariable String projectId, @PathVariable String id) {
        return ResponseEntity.ok(releaseService.publish(id, AuthContext.userId()));
    }

    @PostMapping("/{id}/gates/{gateName}/exempt")
    public ResponseEntity<GateExemption> exempt(@PathVariable String projectId,
                                                 @PathVariable String id,
                                                 @PathVariable String gateName,
                                                 @Valid @RequestBody GateExemptionRequest request) {
        return ResponseEntity.ok(releaseService.exemptGate(id, gateName, request, AuthContext.userId()));
    }

    @GetMapping("/{id}/exemptions")
    public ResponseEntity<List<GateExemption>> exemptions(@PathVariable String projectId,
                                                           @PathVariable String id) {
        return ResponseEntity.ok(releaseService.getExemptions(id));
    }
}

File: services/api/src/release/release.controller.ts
import { Controller, Get, Post, Put, Param, Body } from '@nestjs/common';
import { ReleaseService } from './release.service';
import { CurrentUser, AuthPrincipal } from '../common/decorators/current-user.decorator';
import { ProjectResourceAccessPolicy } from '../common/access/project-resource-access-policy';
import { CreateReleaseDto } from './dto/create-release.dto';
import { ExemptGateDto } from './dto';

/**
 * 发布（Release）Controller。
 *
 * 关键点：
 * - 所有写入均使用 class-validator DTO（KI-005）。
 * - 子资源 ID 走统一 ProjectResourceAccessPolicy 反查归属（KI-004 / R0-04）。
 * - 豁免 Gate 必须提供 reason，写入审计（KI-020 修复要求）。
 */
@Controller('projects/:projectId/releases')
export class ReleaseController {
  constructor(
    private readonly releaseService: ReleaseService,
    private readonly accessPolicy: ProjectResourceAccessPolicy,
  ) {}

  @Post()
  async create(
    @Param('projectId') projectId: string,
    @Body() body: CreateReleaseDto,
    @CurrentUser() user: AuthPrincipal,
  ) {
    await this.accessPolicy.requireProject(projectId, user.userId, user.workspaceId!, 'DEVELOPER');
    return this.releaseService.createRelease(projectId, body, user.userId);
  }

  @Get()
  async list(@Param('projectId') projectId: string, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireProject(projectId, user.userId, user.workspaceId!, 'VIEWER');
    return this.releaseService.listReleases(projectId);
  }

  @Get(':id')
  async get(@Param('id') id: string, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireRelease(id, user.userId, user.workspaceId!, 'VIEWER');
    return this.releaseService.getRelease(id);
  }

  @Get(':id/gates')
  async gates(@Param('id') id: string, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireRelease(id, user.userId, user.workspaceId!, 'VIEWER');
    return this.releaseService.getGateStatus(id);
  }

  @Put(':id/publish')
  async publish(@Param('id') id: string, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireRelease(id, user.userId, user.workspaceId!, 'ADMIN');
    return this.releaseService.publish(id, user.userId);
  }

  @Post(':id/gates/:gateName/exempt')
  async exempt(
    @Param('id') id: string,
    @Param('gateName') gateName: string,
    @Body() body: ExemptGateDto,
    @CurrentUser() user: AuthPrincipal,
  ) {
    await this.accessPolicy.requireRelease(id, user.userId, user.workspaceId!, 'ADMIN');
    return this.releaseService.exemptGate(id, gateName, body, user.userId);
  }

  @Get(':id/exemptions')
  async exemptions(@Param('id') id: string, @CurrentUser() user: AuthPrincipal) {
    await this.accessPolicy.requireRelease(id, user.userId, user.workspaceId!, 'VIEWER');
    return this.releaseService.getExemptions(id);
  }
}
